### PR TITLE
Add agent type

### PR DESCRIPTION
**Pull request by rf-ivtdai**
client-side update to accommodate adding agent_type to agent attributes

### CODE REVIEW

File: invertedai/common.py
Comment (Ruishenl) [inline invertedai/common.py:148]: Can you make the type here the Enum instead of str ?

Comment (adscib) [reply invertedai/common.py:148]: I'm actually not sure that an enum is preferable. It would force people to update the package whenever we introduce additional types.

File: invertedai/common.py
Comment (adscib) [inline invertedai/common.py:52]: `abs` is redundant when taking a square.

File: invertedai/common.py
Comment (adscib) [inline invertedai/common.py:156]: If you're asserting that `l` only has one element, `agent_type, = l` would be preferable. If it can have more, I'm a little worried about how you know that the agent type comes first.

File: invertedai/api/validations.py
Comment (adscib) [inline invertedai/api/validations.py:17]: Is it actually illegal to use `agent_type=None` here?

File: invertedai/api/validations.py
Comment (adscib) [inline invertedai/api/validations.py:24]: This is not allowed under the type signature.

File: invertedai/common.py
Comment (Ruishenl) [reply invertedai/common.py:148]: If we are using string enums it will be automatically casted as string and when we introduce new types we just append more values. Plus it is  using enums for validation already on client side

File: invertedai/api/validations.py
Comment (adscib) [inline invertedai/api/validations.py:35]: I guess "not smaller"?

File: invertedai/api/validations.py
Comment (adscib) [inline invertedai/api/validations.py:28]: Should we also check that all lists in states_history have the same length?

File: invertedai/api/validations.py
Comment (adscib) [inline invertedai/api/validations.py:62]: Here and throughout, the more information you can include in the error message, the better. For example, you can specify the lengths of both lists, and the first unsupported agent type found.

**Review by adscib — CHANGES_REQUESTED** [review]
Generally looks good. I left some comments to make sure that the validators match the intended semantics.

We should also update documentation throughout. At the very least, update the user guide to indicate that we support pedestrians, and the main docstring for DRIVE to warn that it doesn't currently support pedestrians, but will in the future.

File: invertedai/api/validations.py
Comment (Ruishenl) [inline invertedai/api/validations.py:40]: We should mention pedestrian doesn't have lr in the error messages here

File: invertedai/api/validations.py
Comment (Ruishenl) [inline invertedai/api/validations.py:12]: Maybe we should raise a type not supported exception here ?

File: invertedai/common.py
Comment (Ruishenl) [inline invertedai/common.py:156]: It's probably safer to do agent_type = l[-1]

**Comment by rf-ivtdai** [comment]
I have addressed the comments and documented the new features in API and class docstrings. I am a bit unsure as to whether we want to keep the client-side request validations for initialize and drive. After testing, the current checks are all handled gracefully by the server. Replicating the same logic here is redundant and error-prone.

File: invertedai/common.py
Comment (rf-ivtdai) [reply invertedai/common.py:148]: keeping string as per our conversation.

File: invertedai/common.py
Comment (adscib) [inline invertedai/common.py:148]:
```suggestion
    agent_type: Optional[str] = 'car'  #: Valid types are those in `AgentType`, but we use `str` here for extensibility.
```
I actually liked having an `AgentType` enum, even though we use `str` here. We need to have an authoritative list of valid agent types somewhere, and it only makes sense that we would expose it programmatically. Let's just be clear that this enum is extensible and we expect it to grow in future releases.

**Review by adscib — COMMENTED** [review]
I updated the docstring, but please make sure it is accurate.

**Comment by rf-ivtdai** [comment]
To better measure alignments between client and server, i'll add regression tests that mirrors the server integration tests for initialize and drive

**Comment by KieranRatcliffeInvertedAI** [comment]
Changes to the C++ SDK are complete, @adscib is there anything else that needs to be changed before we can merge?

**Comment by rf-ivtdai** [comment]
Regression tests added:
initialize: 10 positive flow - expect 200
                  5 negative flow - expect 400
drive: 4 positive flow - expect 200
drive: 1 negative flow - expect 400
These are all relevant to add_agent_types feature, all passing against local server

**Comment by KieranRatcliffeInvertedAI** [comment]
@adscib I know @rf-ivtdai has been talking to me about adding regression tests which he seems to have added some on the python SDK. Should we add some to C++ SDK or should that be a different PR?

**Comment by adscib** [comment]
It would be nice to add some tests, but I don't want to delay this release, so a separate PR please. Are we waiting for anything here, or can we release already?

**Comment by rf-ivtdai** [comment]
We are ready to release!

**Pull request merged**
